lib/utils/helpers.coffee

```diff
@@ -93,10 +93,25 @@ exports.osProgressHandler = (step) ->
 	return rindle.wait(step)
 
 exports.getAppInfo = (application) ->
+	# Check for an app of the form `user/application`, and send
+	# this off to a special handler (before importing any modules)
+	if (match = /(\w+)\/(\w+)/.exec(application))
+		return getAppInfoWithUser(match[1], match[2])
+
+	resin = require('resin-sdk-preconfigured')
+	return getAppInfoFromPromise(resin.models.application.get(application))
+
+exports.getAppInfoWithUser = getAppInfoWithUser = (username, appName) ->
+	resin = require('resin-sdk-preconfigured')
+	return getAppInfoFromPromise(resin.models.application.getAppWithOwner(appName, username))
+
+# Helper function to setup application info from a promise that resolves
+# to an application
+getAppInfoFromPromise = (appPromise) ->
 	resin = require('resin-sdk-preconfigured')
 	_ = require('lodash')
 	Promise.join(
-		resin.models.application.get(application),
+		appPromise,
```

Comment:
If you flip this refactoring on its head, it'll be much nicer.

Here you've pulled out the bit that varies as a wrapper, around the function's main logic. Instead, you can push down the bit that varies, and keep the core important logic as the top level. That looks something like:

```coffeescript
exports.getAppInfo = (application) ->
	resin = require('resin-sdk-preconfigured')
	_ = require('lodash')
	Promise.join(
		getApplication(application),
		resin.models.config.getDeviceTypes(),
		(app, config) ->
			config = _.find(config, 'slug': app.device_type)
			if !config?
				throw new Error('Could not read application information!')
			app.arch = config.arch
			return app
	)

getApplication = (application) ->
	resin = require('resin-sdk-preconfigured')

	# Check for an app of the form `user/application`, and send
	# this off to a special handler
	if (match = /(\w+)\/(\w+)/.exec(application))
		return resin.models.application.getAppWithOwner(match[1], match[2])

	return resin.models.application.get(application)
```

Easier to follow the structure that way, and you end up with more composable standalone parts, rather than tying everything together. Also strictly necessary if you have more than one refactoring like this, or you get an explosion of interlinked entry points. Eventually, `getApplication` might be something we even move into the SDK, which is super easy this way around.

It looks like you're also exporting `getAppInfoWithUser`, because you want to explicitly use that elsewhere? If so, you could extend the above to include an optional user argument, pass it to `getApplication`, and handle that there to get the same result. If not, that's fine too :smiley: